### ISSUE TITLE
Wayland resize issue/buffer not freed upon resize

### ISSUE THREAD

**Issue by fcipaq** [libhybris/libhybris#444]
Resize issue/buffer not freed upon resize

I am using libhybris on a x86 Android Tablet to run Plasma/Wayland-Desktop. The Kernel version is 3.10 and I am on Debian testing. Graphics is PowerVR Rogue.

Most basic tasks work well and the desktop is accelerated. Using Firefox 72 w/ Wayland support works well - both in terms of performance and stability.

However, resizing EGL windows does not work properly. I had to change https://github.com/libhybris/libhybris/blob/master/hybris/egl/platforms/wayland/wayland_window.cpp#L92 and L93 to 

```
    this->m_width = width;
    this->m_height = height;
```

in order to make it work at all. I am also unsure about `attached_width` and `attached_height` set in `finishSwap()`

However, resizing Firefox's window works once, but on the second resize event the window locks up.

`HYBRIS_TRACE` reveals that `dequeueBuffer` waits for a free buffer - forever.

After the first resize event one buffer is permanently occupied, as can be seen from the log:

https://paste.ubuntu.com/p/GbhT8tnR8s/

Full log (w/ HYBRIS_TRACE):

https://paste.ubuntu.com/p/vs3yMYwNjc/

(lines beginning with FCIPAQ are some additional infos to make tracking (hopefully) easier).

Any input highly appreciated and thanks for any help in advance.

**Comment by fcipaq** [comment]
I have investigated this a little further an to my understanding the issue comes down to the following:

When a window is resized, the compositor (i.e. Kwin) sends an event to both the EGL driver as well as the client (is that correct?).

Now, the EGL driver applies the new size and any newly dequeued buffer will have new dimensions. The client then requests a new buffer to rerender. The EGL driver then dequeues a buffer (with the new size) and hands it over to the client. The client does its rendering and eventually calls eglSwapBuffers(WithDamage). The EGL driver then enqueues the buffer and finally swaps back und front buffer.

However in libhybris/Kwin it's different. Once the window has been resized, the client receives the resize event and requests a new buffer. BUT libhybris has not yet received a resize callback and dequeues a buffer using the OLD dimensions. Only after the buffer has already been dequeued, libhybris receives a resize event:

```
finishSwap: attaching wlsurface (0xc4c01910): 1136 x 1030

--> resizing window

dequeue 0xd4fe767c START FREEBUFFERS: 1
dequeue 0xd4fe767c END

resize event: old 1136 x 1030, new: 1123 x 1218, 

queue 0xc5055280 start
queue 0xc5055280 END

finishSwap: attaching wlsurface (0xc4c01e20): 1136 x 1030

dequeue 0xd4fe755c START FREEBUFFERS: 0
```

Now, (I guess) there's two things the client can do

a) render anyways to the incorrectly sized buffer and the result will be graphical glitches (like Plasmashell does). Looks like this:

![2](https://user-images.githubusercontent.com/45269484/80147350-aad5cd00-85b3-11ea-8a2e-e47667c96498.png)

b) Reject the buffer all along and request another one. However this causes the already dequeued buffer to be permanently unusable as the Android EGL implementation neither calls `releaseBuffer` nor `cancelBuffer`. Since 3 is the default number of buffers, you can successfully resize the window once before you run out of back buffers and `dequeueBuffer` locks up waiting for a buffer ad infinitum.

tl;dr The issue seems to be that libhybris receives the resize_callback always after the client receives the resize event and an (incorrectly sized buffer) has already been dequeued and the client has already started rendering (not a race condition, checked with a long delay) .

What is the reason for this behavior and how could it possibly be resolved?

Thanks for any input in advance

**Comment by fcipaq** [comment]
Second follow up - third time's a charm :-P

I think that I now might understand what is the actual problem here is:

When a compoitor or a client issues a resize, the client informs EGL to update the EGL _SURFACE_ Firefox for example will reset the context like this: https://github.com/mozilla/gecko-dev/blob/master/gfx/layers/opengl/CompositorOGL.cpp#L980

Now, this call goes to the Android implementation (eglMakeCurrent is just passed through) and is out of scope for hybris currently. The client then immediately prepares to rerender and requests a new buffer (and libhybris gets a `deqeueBuffer` call) and then the client AFTERWARDS sends a resize request to the window system in order to resize the _WINDOW_.
However in libhybris' `wayland_window.cpp` the buffer size is actually the _WINDOW_ size (not the _SURFACE_ size) - which does not seem correct to me but does not matter as long as you don't resize or use a surface which is not the same size as the window. BUT whereas the _SURFACE_ has already been resized, the _WINDOW_ still has the old size until libhybris receives the `resize_callback` which is only after the buffer has already been dequeued and the client has already begun rendering.
So what we actually need here is a callback from inside the Android EGL whenever a surface is resized so that the native window can be resized accordingly.

Mesa has had the inverse problem but as they dont wrap around a propriertary EGL implementation they can simply update the buffer  (https://gitlab.freedesktop.org/FireBurn/mesa/commit/a9fb331ea7d1a78936ea8d8385e44cfd66f835c1#b91f29c819033a507b2cac556bea4f5aecd8893d)

However there might be a solution to the problem: 

a) First suggestion would be to call `eglQuerySurface ` from within libhybris' `dequeueBuffer` to see if the surface size still matches the buffer the size.

b) Second option could be to intercept a call to `eglMakeCurrent` in libhybris to check whether the surface has been resized and in case it does not give the wayland window system a `resize_callback`.

Does this make sense? Any input highly appreciated and thanks in advance.
